<p style="text-align:right;">
    <em>Content Copyrighted by Puneett Bhatnagr</em>
</p>


# String Formatting
___

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. 

As a quick comparison, Lets consider a scenario, where we need to write a string as 'Last night <b>player</b> scored <b>xx</b> points.'

Here we will take <b>player</b> as 'Virat' & <b>points</b> as 36

### In concatenation, we will write

'Last night, '+player+' scored '+str(points)+' points.'  

### In string formatting, we will write

f'Last night, {player} scored {points} points.'
__

There are three ways to perform a string formatting.

1. Formatting with placeholders - The oldest method involves placeholders using the modulo `%` character.

2. An improved technique uses the `.format()` string method.

3. The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

Since you will likely encounter all three versions in someone else's code, we describe each of them in our topic.

# Formatting with placeholders
___

You can use <code>%s</code> to inject strings into your print statements. 

The modulo `%` is referred to as a "string formatting operator".

In [2]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [4]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [5]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


## Format conversion methods.
___

It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. 

We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

%s operator lets you add a value into a Python string.

%r is the Python “representation” for the object, a string which, if presented to a Python interpreter should be parsed as a literal or as an instantiation of a new object of that time and with the same value. 

The %r and %s for many objects is identical.

repr() prints “official” representation of a date-time object (means using the “official” string representation we can reconstruct the object).


In [22]:
print('He said his name was %s.' %'Python') # %s is used to print string without quotes
print('He said his name was %r.' %'Python') # %r is used to print as it is

He said his name was Python.
He said his name was 'Python'.


As another example, `\t` inserts a tab into a string.

In Python strings, the backslash "\" is a special character, also called the "escape" character. It is used in representing certain whitespace characters: "\t" is a tab, "\n" is a newline, and "\r" is a carriage return.

In [23]:
print('I once caught a chicken %s.' %'this \tbig')
print('I once caught a chicken %r.' %'this \tbig')

I once caught a chicken this 	big.
I once caught a chicken 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [24]:
print('I wrote %s programs today.' %2.25)
print('I wrote %d programs today.' %2.25)   

I wrote 2.25 programs today.
I wrote 2 programs today.


## Padding and Precision of Floating Point Numbers
___


Floating point numbers use the format <code>%5.2f</code>. 

Here, <code>5</code> would be the minimum number of characters the string should contain; 

these may be padded with whitespace if the entire number does not have this many digits. 

Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. 

Let's see some examples:

In [1]:
print('Floating point numbers: %5.2f' %(10.166))

Floating point numbers: 10.17


In [2]:
print('Floating point numbers: %1.0f' %(10.166))

Floating point numbers: 10


In [3]:
print('Floating point numbers: %1.5f' %(10.166))

Floating point numbers: 10.16600


In [4]:
print('Floating point numbers: %10.2f' %(10.166))

Floating point numbers:      10.17


In [5]:
print('Floating point numbers: %25.2f' %(10.166))

Floating point numbers:                     10.17


For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

# Multiple Formatting
___


Nothing prohibits using more than one conversion tool in the same print statement:

In [30]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',1.1615,'bye!'))

First: hi!, Second:  1.16, Third: 'bye!'


## Formatting with the `.format()` method
___


A better way to format objects into your strings for print statements is with the string `.format()` method. 

The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [3]:
print('This is a string {}'.format('inserted'))

This is a string inserted


## .format() method
___


The .format() method has several advantages over the %s placeholder method:

#### 1. Inserted objects can be called by index position:

In [32]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


#### 2. Inserted objects can be assigned keywords:

In [33]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Inserted objects can be reused, avoiding duplication:

In [35]:
print('A %s saved is a %s earned.' %('rupee','rupee'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='rupee'))

A rupee saved is a rupee earned.
A rupee saved is a rupee earned.


### Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [10]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [11]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [12]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [13]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

## Formatted String Literals (f-strings)
___



f-strings offer several benefits over the older `.format()` string method described above. 

For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [14]:
name = 'Virat'

print(f"He said his name is {name}.")

He said his name is Virat.


Pass `!r` to get the string representation:

In [15]:
print(f"He said his name is {name!r}")

He said his name is 'Virat'


### Float formatting follows `"result: {value:{width}.{precision}}"`

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [16]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. 

This fits more closely with scientific notation and statistical analysis. 

Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [17]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [18]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


For more info on formatted string literals visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

# Let's Summarise what we learn
___


So what we learned in this Topic -

- String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation.

- Convert any python object to a string using two separate methods: `str()` and `repr()`.

- Floating point numbers use the format <code>%5.2f</code>.

- format objects into your strings for print statements is with the string `.format()` method.



# Lets now move to next Topic - Lists!

___

<p style="text-align:right;">
    <em>Content Copyrighted by Puneett Bhatnagr</em>
</p>